File: src/main/java/es/uam/eps/ir/recsys/datasets/data/Impressions.java
/*
 * Copyright (C) 2021 Information Retrieval Group at Universidad Autónoma
 * de Madrid, http://ir.ii.uam.es.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
package es.uam.eps.ir.recsys.datasets.data;

import java.util.*;
import java.util.stream.Stream;

/**
 * Class that represents the basic rating matrix of a recommendation dataset.
 * It just includes: a) users b) items c) ratings between users and items.
 *
 * @author Javier Sanz-Cruzado (javier.sanz-cruzado@uam.es)
 */
public class Impressions
{
    /**
     * The number of ratings in the system.
     */
    private int numImpressions;

    /**
     * User-to-item rating matrix.
     */
    private final Map<Integer, Set<Integer>> userImpressions;
    /**
     * Item-to-user rating matrix.
     */
    private final Map<Integer, Set<Integer>> itemImpressions;

    /**
     * Constructor.
     */
    public Impressions()
    {
        // Initialize the structures:
        this.userImpressions = new HashMap<>();
        this.itemImpressions = new HashMap<>();
    }

    /**
     * We add a new user to the matrix.
     * @param userId the identifier of the user.
     * @return true if we add it, false otherwise (if it exists).
     */
    public boolean addUser(int userId)
    {
        if(userImpressions.containsKey(userId)) return false;
        this.userImpressions.put(userId, new HashSet<>());
        return true;
    }

    /**
     * We add a new item to the matrix.
     * @param itemId the identifier of the item.
     * @return true if we add it, false otherwise (if it exists).
     */
    public boolean addItem(int itemId)
    {
        if(itemImpressions.containsKey(itemId)) return false;
        this.itemImpressions.put(itemId, new HashSet<>());
        return true;
    }

    /**
     * We add a new impression to the matrix.
     * @param userId the identifier of the user.
     * @param itemId the identifier of the item.
     * @return true if we add it or we update a previous case, false otherwise (if it exists).
     */
    public AddingReturn addImpression(int userId, int itemId)
    {
        if(userImpressions.containsKey(userId) && itemImpressions.containsKey(itemId))
        {
            if (userImpressions.get(userId).contains(itemId))
            {
                return AddingReturn.NONE;
            }
            else
            {
                userImpressions.get(userId).add(itemId);
                itemImpressions.get(itemId).add(userId);
                numImpressions++;
                return AddingReturn.ADDED;
            }
        }

        return AddingReturn.NONE;
    }

    /**
     * Obtains the set of users.
     * @return the set of users.
     */
    public Stream<Integer> getUsers()
    {
        return this.userImpressions.keySet().stream();
    }

    /**
     * Obtains the set of items.
     * @return the set of items.
     */
    public Stream<Integer> getItems()
    {
        return this.itemImpressions.keySet().stream();
    }

    /**
     * Obtains the impressions of a user.
     * @param userId the identifier of the user.
     * @return the impressions of the user (an empty stream if there are none, or the user does not even exist).
     */
    public Stream<Integer> getUserImpressions(int userId)
    {
        if(!userImpressions.containsKey(userId)) return Stream.empty();
        return userImpressions.get(userId).stream();
    }

    /**
     * Obtains the ratings of a item.
     * @param itemId the identifier of the item.
     * @return the ratings of the item (an empty stream if there are none, or the item does not even exist).
     */
    public Stream<Integer> getItemImpressions(int itemId)
    {
        if(!itemImpressions.containsKey(itemId)) return Stream.empty();
        return itemImpressions.get(itemId).stream();
    }

    /**
     * Obtains the number of impressions.
     * @return the number of impressions.
     */
    public int getNumImpressions()
    {
        return this.numImpressions;
    }

    public int getNumUsers()
    {
        return this.userImpressions.size();
    }

    public int getNumItems()
    {
        return this.itemImpressions.size();
    }
}

File: src/main/python/data/Impressions.py
"""
Representation of the impressions included in a dataset.
"""

__version__ = '0.1'
__author__ = 'Javier Sanz-Cruzado, Pablo Castells'
__email__ = 'javier.sanz-cruzado@uam.es, pablo.castells@uam.es'

from data import AddingReturn


class Impressions:
    """
    Class that represents the impressions, i.e. the set of items which have been shown to a user.
    """

    def __init__(self):
        """
        Initializes the impressions in the system.
        """
        self.user_impressions = dict()
        self.item_impressions = dict()
        self.num_impressions = 0

    def add_user(self, user_id):
        """
        Adds a new user to the rating matrix.
        :param user_id: the identifier of the user.
        :return: true if we add the user, false otherwise.
        """
        if self.user_impressions.keys().__contains__(user_id):
            return False
        else:
            self.user_impressions[user_id] = set()
            return True

    def add_item(self, item_id):
        """
        Adds a new item to the rating matrix
        :param item_id: the identifier of the item.
        :return: true if we add the item, false otherwise.
        """
        if self.item_impressions.keys().__contains__(item_id):
            return False
        else:
            self.item_impressions[item_id] = set()
            return True

    def add_impression(self, user_id, item_id):
        """
        Adds a new impression
        :param user_id: the identifier of the user.
        :param item_id: the identifier of the item.
        :return: AddingReturn.ADDED if the impression is new, AddingReturn.NONE if it already existed,
                 AddingReturn.ERROR if something failed.
        """

        if self.user_impressions.keys().__contains__(user_id) and self.item_impressions.keys().__contains__(item_id):
            if self.user_impressions[user_id].__contains__(item_id):
                return AddingReturn.NONE
            else:
                self.user_impressions[user_id].add(item_id)
                self.item_impressions[item_id].add(user_id)
                self.num_impressions += 1
                return AddingReturn.ADDED
        return AddingReturn.ERROR

    def get_users(self):
        """
        Obtains the set of users.
        :return: the set of users.
        """
        return iter(self.user_impressions.keys())

    def get_items(self):
        """
        Obtains the set of items.
        :return: the set of items.
        """
        return iter(self.item_impressions.keys())

    def get_user_impressions(self, user_id):
        """
        Obtains the set of impressions for a user.
        :param user_id: the identifier of the user.
        :return: an iterator of the impressions for the user.
        """
        return iter(self.user_impressions.get(user_id, set()))

    def get_item_impressions(self, item_id):
        """
        Obtains the set of users including an item in their impressions.
        :param item_id: the identifier of the item.
        :return: an iterator of the impressed users.
        """
        return iter(self.item_impressions.get(item_id, set()))

    def get_num_impressions(self):
        """
        Obtains the number of impressions.
        :return: the number of impressions.
        """
        return self.num_impressions

    def get_num_users(self):
        """
        Obtains the number of users.
        :return: the number of users.
        """
        return self.user_impressions.keys().__len__()

    def get_num_items(self):
        """
        Obtains the number of items.
        :return: the number of items.
        """
        return self.item_impressions.keys().__len__()